internal/campaigns/run_steps.go

```diff
@@ -297,3 +308,57 @@ func probeImageForShell(ctx context.Context, image string) (shell, tempfile stri
 	// admit defeat and return. At least err is already in place.
 	return
 }
+
+type StepFailedErr struct {
+	Run       string
+	Container string
+
+	TmpFilename string
+
+	Args   []string
+	Stdout string
+	Stderr string
+
+	Err error
+}
+
+func (e StepFailedErr) Error() string {
+	var out strings.Builder
+
+	fmtRun := func(run string) string {
+		lines := strings.Split(run, "\n")
+		if len(lines) == 1 {
+			return lines[0]
+		}
+		return lines[0] + fmt.Sprintf("\n\t(... and %d more lines)", len(lines)-1)
+	}
+
+	out.WriteString(fmt.Sprintf("run: %s\ncontainer: %s\n", fmtRun(e.Run), e.Container))
+
+	printOutput := func(output string) {
+		for _, line := range strings.Split(output, "\n") {
+			if e.TmpFilename != "" {
+				line = strings.ReplaceAll(line, e.TmpFilename+": ", "")
+			}
+			out.WriteString("\t" + line + "\n")
+		}
+	}
+
+	if len(e.Stdout) > 0 {
+		out.WriteString("\nstandard out:\n")
+		printOutput(e.Stdout)
+	}
+
+	if len(e.Stderr) > 0 {
+		out.WriteString("\nstandard error:\n")
+		printOutput(e.Stderr)
+	}
```

Comment:
Do you think they should be printed separately? I don't really know many places where that's the case (CI providers, GH actions, ..) usually print everything together (and sometimes it could be confusing I guess if:

```
stdout: Processing resource file X.yml
stderr: Failed to process.
stdout: Processing resource file Y.yml
stdout: Done, 1 errors.
```
, how would I know which of the files failed. Maybe it's just me though, so leaving it up to you and approve
)

Reply:
Yeah, they print everything together because they stream the output, we don't. We buffer and then print. And since we buffer them separately we can't easily weave them back together to get to your example output. We'd need to record some timing information for that, but that's a hole I don't want to jump into. Or we could use a custom entity that buffers and prefixes both.

Honestly, though, I like having them separate. I hate having to dig through lines and lines of output to get to the error messages. It's something that annoys me about CI providers actually 😄

Reply:
we could use the same buffer though, no? 🤔 I mean, which tool prints simultaneously on both out and err 🤔 

But if you like it better, go for it, I was just scared that some tools might use the channels in a weird way that only works when they're printed next to each other, like in my sample.
Probably an unreasonable fear though 😬

Reply:
Yeah, sure, we could use the same buffer, but then we'd just have everything together in one without knowing at a quick glance what's error or output. And considering how noisy some tools are... If we run into issues, we can revisit this, I think.

Reply:
totally